packages/compiler-core/src/transforms/vFor.ts

```diff
@@ -206,7 +208,40 @@ export const transformFor: NodeTransform = createStructuralDirectiveTransform(
             helper(getVNodeHelper(context.inSSR, childBlock.isComponent))
           }
         }
+        if (__DEV__ || !__BROWSER__) {
+          let currentTag = ''
+          if (forNode.children[0]) {
+            currentTag = (forNode.children[0] as BaseElementNode).tag
+          }
+          if (
+            currentTag &&
+            forNode.parseResult.value &&
+            (forNode.parseResult.value as SimpleExpressionNode).content ===
+              currentTag
+          ) {
+            context.onError(
+              createCompilerError(ErrorCodes.X_V_FOR_PARAMS, childBlock.loc),
+            )
+          }
+          let identifiers: CompoundExpressionNode['identifiers'] = []
+          if (
+            forNode.parseResult.value &&
+            (forNode.parseResult.value as CompoundExpressionNode).identifiers
+          ) {
+            identifiers = (forNode.parseResult.value as CompoundExpressionNode)
+              .identifiers
+          }
 
+          if (
+            currentTag &&
+            identifiers &&
+            identifiers.some(i => i === currentTag)
```

Comment:
This check is not correct. Consider the following scenario:
```vue
  <div v-for="Comp of list">
    <Comp>{{Comp}}</Comp>
  </div>
```
[Playground](https://play.vuejs.org/#__PROD__eNp9UrFOwzAQ/RXLS5eQqJSppJUAdYABELBhhiq9FBfHtmwnRIry79w5SolEyxI79949v/Nzx2+sTZsa+JLnvnDSBuYh1HYttKyscYHdmcqy0pmKzdKMfog+uxZa6MJoH5iS+Fmx93nCLhO2+CAkzwYxlMkDVFZtA+CesXwnG9ZclMatBI/SpowKgkccGVRddx0tfZ/HI4fWDHtJMJso8oQHjz5KuU8P3micoyOy4AX2SQXuyQaJPgVfsogQtlXKfD/EWnA1JGO9+ITi60T94FuqCf7swINrQPAjFrZuD2if4M3rI7S4P4KV2dUK2f+AL+CNqsnjQLut9Q5tT3jR7X1MQ+r9m9+0AbQfhyKjxOwjX3BMh67s3Oi/dhfpVewTusdbHJP98xAmUZ4Kc0zNKxPWcyTTej6u/ge0D86e)